### ISSUE TITLE
Return type of decorated function not correctly inferred

### ISSUE THREAD

**Issue by SandyChapman** [microsoft/pyright#2688]
**Describe the bug**
Consider the following code:

```python
def adds_foo_1(param: str):
    return param + "foo"

adds_foo_1() # Argument missing for parameter "param"
```

Pyright correctly infers the type of the function to require a string param. However, if we decorate this function, the type inference is incorrect:

```python
from functools import wraps

def foo_caller(fn):
    return fn('hardcoded')

def wrap_fn(fn):
    @wraps(fn)
    def _inner(*args, **kwargs):
        return foo_caller(fn)
    return _inner

@wrap_fn
def adds_foo_2(param: str):
    return param + "foo"

print(adds_foo_2()) # Argument missing for parameter "param"
# Outputs "hardcodedfoo" at runtime
```

In this code, the return type of `adds_foo_2` is substituted with a callable that no longer requires the string param to be specified. Pyright should catch this and not result in an error.

Lastly,
```
def adds_foo_3(param: str):
    return param + "foo"

wrapped_adds_foo_3 = wrap_fn(adds_foo_3)
print(wrapped_adds_foo_3()) # No error here.
# Outputs "hardcodedfoo"
```
Manually wrapping the function results in the decorated function type being correctly resolved.

**To Reproduce**
Run the above code with pyright's basic type checking mode.

**Expected behavior**
Pyright should resolve the decorated function's type.

**Screenshots or Code**

![image](https://user-images.githubusercontent.com/3920249/145472047-07830bfe-722f-436e-9200-35c91e377c3b.png)

**VS Code extension or command-line**

➤ pyright --version
pyright 1.1.194

Also, Visual Studio Code version 1.62.3 w/ Pylance v2021.12.0

**Comment by erictraut** [comment]
You need to add type annotations in this case. When a decorator is unannotated, pyright (and other type checkers such as mypy) assume that the decorator leaves the signature of the decorated function unchanged. In most cases, this is the right assumption to make. If the decorator mutates the signature, then an explicit annotation is needed.

**Comment by SandyChapman** [comment]
Ah perfect! So it does. Thanks!
